hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdb/util/MinHeapMergeIterator.java

```diff
@@ -0,0 +1,211 @@
+    this.initialized = false;
+    this.comparator = comparator;
+  }
+
+  protected abstract I getIterator(int idx) throws RocksDBException, IOException;
+
+  private boolean initHeap() throws IOException, RocksDBException {
+    if (initialized) {
+      return false;
+    }
+    initialized = true;
+    int count = 0;
+    try {
+      for (int idx = 0; idx < iterators.size(); idx++) {
+        I itr = getIterator(idx);
+        iterators.set(idx, itr);
+        HeapEntry<K> entry = new HeapEntry<>(idx, itr, comparator);
+        if (entry.getCurrentKey() != null) {
+          minHeap.add(entry);
+          count++;
+        } else {
+          // No valid entries, close the iterator.
+          closeItrAtIndex(idx);
+        }
+      }
+    } catch (IOException | RocksDBException e) {
+      close();
+      throw e;
+    }
+    return count > 0;
+  }
+
+  @Override
+  public boolean hasNext() {
+    try {
+      return !minHeap.isEmpty() || initHeap();
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    } catch (RocksDBException e) {
+      throw new UncheckedIOException(new RocksDatabaseException("Error while initializing iterator ", e));
+    }
+  }
+
+  protected abstract V merge(Map<Integer, K> keysToMerge);
+
+  @Override
+  public V next() {
+    if (!hasNext()) {
+      throw new NoSuchElementException("No more elements found.");
+    }
+
+    assert minHeap.peek() != null;
+    // Get current key from heap
+    K currentKey = minHeap.peek().getCurrentKey();
+    // Clear the keys list by setting all entries to null.
+    keys.clear();
+    // Advance all entries with the same key (from different files)
+    while (!minHeap.isEmpty() && Objects.equals(minHeap.peek().getCurrentKey(), currentKey)) {
+      HeapEntry<K> entry = minHeap.poll();
+      int idx = entry.index;
+      // Set the key for the current entry in the keys list.
+      keys.put(idx, entry.getCurrentKey());
+      if (entry.advance()) {
+        minHeap.offer(entry);
+      } else {
+        // Iterator is exhausted, close it to prevent resource leak
+        try {
+          closeItrAtIndex(idx);
+        } catch (IOException e) {
+          throw new UncheckedIOException(e);
+        }
+      }
+    }
+    return merge(keys);
+  }
+
+  private void closeItrAtIndex(int idx) throws IOException {
+    if (iterators.get(idx) != null) {
+      iterators.get(idx).close();
+      iterators.set(idx, null);
+    }
+  }
+
+  @Override
+  public void close() {
+    IOException exception = null;
+    for (int idx = 0; idx < iterators.size(); idx++) {
+      try {
+        closeItrAtIndex(idx);
+      } catch (IOException e) {
+        exception = e;
+      }
+    }
+    if (exception != null) {
+      throw new UncheckedIOException(exception);
+    }
+  }
+
+  /**
+   * A wrapper class that holds an iterator and its current value for heap operations.
+   */
+  private static final class HeapEntry<T> implements Comparable<HeapEntry<T>> {
```

Comment:
this is moved from SstFileSetRead.HeapEntry, with index and comparator.
index is not used.
comparator is customizable.